src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs

```diff
@@ -377,5 +377,164 @@ public virtual void EnableExtensions(bool enable = true)
             var rc = raw.sqlite3_enable_load_extension(_db, enable ? 1 : 0);
             SqliteException.ThrowExceptionForRC(rc, _db);
         }
+
+        private void CreateFunctionCore<TState, TResult>(
+            string name,
+            int arity,
+            TState state,
+            Func<TState, SqliteValueReader, TResult> function)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (State != ConnectionState.Open)
+            {
+                throw new InvalidOperationException(Resources.CallRequiresOpenConnection(nameof(CreateFunction)));
+            }
+
+            delegate_function_scalar func = null;
+            if (function != null)
+            {
+                func = (ctx, user_data, args) =>
+                {
+                    // TODO: Avoid allocation when niladic
```

Comment:
The TODOs left in this PR are theoretical perf improvements. I didn't want to make them without measuring the improvement since they would complicate the code base.